### PR TITLE
Fix decorator evaluation private environment

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/16295
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

In this PR we move the element decorator evaluation from the preface of class declaration into computed key.

Previously they are evaluated outside the class, immediate after class decorator evaluations. However, because the element decorators are defined within the class syntax, they should have access to the private variables defined in the current class. 

The current evaluation storage:

1) The computed key are the ideal place to store element decorator evaluations because they share the same scope.
2) If a computed key is not available, we will convert the last public key into a computed key.
3) If a class has only private elements and static blocks, we will insert the decorator evaluations before the `applyDecs` call. We will replace all function context usage before moving them into the static block. However, this is still not safe because decorators expressions / computed keys inherit the outer level `[Yield]` and `[Await]` production parameter, while the static block where we insert `applyDecs` call can not be a generator / async function. Given that the scenaraio of private-only class is rare, I would like to trade it with this general private fix.

I suggest review this PR by commits.

### CODE REVIEW

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        @(yield* super.id(_ => _.#X), dummy) #property;
+        @(yield* super.id(_ => _.#X), dummy) static #Property;
+        @(yield* super.id(_ => _.#X), dummy) accessor #accessor;
+        @(yield* super.id(_ => _.#X), dummy) static accessor #Accessor;
+      }
+    }
+  }
+
+  expect([...A].map(fn => fn(B)).join(",")).toBe("B#X,B#X,B#X,B#X,B#X,B#X,B#X,B#X,B#X,B#X");
+}
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js:33]: This edge case test is disabled because I can't figure out how to transform it.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-scope/exec.js [hunk 2]
```diff
@@ -1,12 +0,0 @@
...
-class B {
-  static m() {
-   class C {
-     @decFactory(this.name) p;
-   }
-  }
-}
-
-B.m();
-expect(receivedName).toBe("B");
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-scope/exec.js:12]: This test was renamed to "decorator-evaluation-this" and then augmented.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js (hunk 1 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js:33]: In this case, can we introduce a new public element and delete it immediately after defined?
```js
class C {
  // for demo purpose, the `delete` expression can be directly inserted into staticInitializerAssignments
  @(function(_, context) { context.addInitializer(function() { delete this[temp] } })
  static [/* decorator evaluations */, temp = Symbol()];
}
```
Of course using `delete` is a performance killer, but this approach will be only invoked when there is no public class elements, which should be quite rare.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
-var _initStatic, _init_a, _init_a2, _get_a, _set_a, _init_computedKey, _init_computedKey2, _init_computedKey3, _init_computedKey4, _init_computedKey5, _init_computedKey6, _computedKey, _init_computedKey7, _Foo;
+let _computedKey;
+var _initStatic, _init_a, _init_a2, _get_a, _set_a, _init_computedKey, _init_computedKey2, _init_computedKey3, _init_computedKey4, _init_computedKey5, _init_computedKey6, _init_computedKey7, _Foo;
```
Comment (liuxingbaoyu) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js:2]: Can we revert this change?

**Comment by nicolo-ribaudo** [comment]
"I suggest review this PR by commits." ~ 13 commits 😛

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-super-property/exec.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  expect(receivedName).toBe("A");
+}
+
+{
+  let receivedName;
+  class B extends class A {} {
+    static m() {
+      @noop
+      class C {
+        @(receivedName = super.name, noop) p;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-super-property/exec.js:51]: Can you add a test with a decorator that references both `super` and a private field in the same decorator?

File: packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js (hunk 3 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js:2]: This change is intentional as it will fix an integration issue with the class properties transform. For context see https://github.com/babel/babel/pull/10029/files#r287785963 

In this [REPL example](https://babel.dev/repl#?browsers=chrome%2032&build=&builtIns=false&corejs=3.21&spec=false&loose=true&code_lz=GYVwdgxgLglg9mABGOcAOAKAlIg3gXwCgIEBnKRCAGwENTSBTUxAXkQG0BdAbkODgBOiDFQYUYrRAAZuiCQB5EAVlkBqVTBy5CiSrXpMAdGhCkAFhh269dZgEE8V64gACKdE-vsYnSQCIAWwBPRDQBdAYBKCC_XmciXSxeIgA3GiEBJhAqCjYuXn4hDBIwchsAL3LEOGAbA1ItK0zSbKhjUwsAeQAjACsGaEMAczFOgHcwAAVwtEjogDkaAKYMMAYxivLsLHYpTiTCIhLSOFFDKjghjGbWw164GDBtoA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=&version=7.24.0&externalPlugins=%40babel%2Fplugin-proposal-decorators%407.23.3%2C%40babel%2Fplugin-transform-class-properties%407.23.3%2C%40babel%2Fplugin-transform-class-static-block%407.23.4&assumptions=%7B%7D)

```js
function noop() {}
const classes = [];
for (let i = 0; i < 5; ++i) {
  classes.push(
    class A {
      @noop
      [i] = "my property";
    }
  );
}
var result = [];
for (const clazz of classes) {
  result.push(Object.getOwnPropertyNames(new clazz())[0]);
}
console.log(result.join())
```
when we apply the decorators, static block and class properties plugin, it should print `0,1,2,3,4` but currently it prints `4,4,4,4,4`.

If the output size is the concern, I can open a new PR to register all variable as the `let` binding.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 5]
```diff
@@ -1640,35 +1651,82 @@ function transformClass(
...
+
+  if (computedKeyAssignments.length > 0) {
+    applyDecoratorWrapperPath.pushContainer(
+      "body",
+      t.expressionStatement(maybeSequenceExpression(computedKeyAssignments)),
+    );
+
+    const computedKeysPath = applyDecoratorWrapperPath.get("body")[0];
+
+    // Capture lexical this and super call, replace their usage in computed key assignments
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: Good point! I didn't realize we have such a utility in the traverse library. However I am down to the temporary property approach and hopefully we can remove all the lexical variables capturing logic here.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-super-property/exec.js (hunk 4 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-super-property/exec.js:51]: Does this test look good?

https://github.com/babel/babel/blob/0b8f0bc3acc8ff3f44f0fa6931d6bad3dec9c0a3/packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc--to-es2015/decorator-evaluation-yield-private-super-property/exec.js#L12-L29

File: packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js (hunk 3 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js:2]: Yes please add a test!

Regarding to output size, we should first aim at correctness and _then_ optimization.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-super-property/exec.js (hunk 4 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/decorator-evaluation-super-property/exec.js:51]: This is ok, I somehow missed it.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js (hunk 3 above)
Comment (liuxingbaoyu) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js:2]: Thanks! I didn't notice this was a bug fix, I thought the increased size didn't matter.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js (hunk 1 above)
Comment (liuxingbaoyu) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js:33]: That sounds good to me.
We can even do it when someone needs it. :)

**Comment by liuxingbaoyu** [comment]
test262 seems to be relevant, while the other CI failures are not.

Considering the complexity of rebasing this PR, we can postpone the merge of https://github.com/babel/babel/pull/16326.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 6]
```diff
@@ -277,6 +278,84 @@ function extractProxyAccessorsFor(
...
+}
+
+function prependExpressionsToComputedKey(
+  expressions: t.Expression[],
+  fieldPath: NodePath<
+    t.ClassMethod | t.ClassProperty | t.ClassAccessorProperty
+  >,
+) {
+  const key = fieldPath.get("key") as NodePath<t.Expression>;
+  expressions.push(key.node);
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts:318]: If a `key` is already a sequence expression, it could be generated by another Babel plugin. The intention here is try to reuse the uid memoiser, for example the computed keys transform can reuse the memoiser `_computedKeys` generated by the decorator transform:

```js
class C {
  [decs = f(), _computedKey = g(), another_decs = h(), _computedKey];
}
```

but at this point there is no class features running before decorators, so there aren't much practical examples. I will add a test case for nested sequence expression.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js (hunk 1 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js:33]: > for delete, let's put it in a static { delete this._ } block at the beginning of the class?

I plan to insert the delete expression right before the `applyDecs` call, since its parent block is already the first class element.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js (hunk 3 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-accessors--to-es2015/context-name/output.js:2]: 2cd5bcc2007b8e9fea907bc96eb5606334f038a3

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: 6969fd41db1aef76a8c0e63e5551ee2bb025d592

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 7]
```diff
@@ -277,6 +278,84 @@ function extractProxyAccessorsFor(
...
+      } else {
+        completion.replaceWith(
+          maybeSequenceExpression([
+            t.cloneNode(maybeAssignment),
+            ...expressionSequence,
+          ]),
+        );
+      }
+    } else {
+      // If the computed key is an uid reference, treat it as
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: 6969fd41db1aef76a8c0e63e5551ee2bb025d592

File: packages/babel-helper-create-class-features-plugin/src/misc.ts [hunk 8]
```diff
@@ -120,14 +120,48 @@ export function injectInitialization(
...
+  scope: Scope,
+  hint: string,
+): ComputedKeyAssignmentExpression | undefined {
+  const isUidReference = t.isIdentifier(keyNode) && scope.hasUid(keyNode.name);
+  const isMemoiseAssignment =
+    t.isAssignmentExpression(keyNode, { operator: "=" }) &&
+    t.isIdentifier(keyNode.left) &&
+    scope.hasUid(keyNode.left.name);
+  if (isUidReference) {
+    return;
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/misc.ts]: 6969fd41db1aef76a8c0e63e5551ee2bb025d592

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: https://github.com/babel/babel/pull/16325/commits/f4314e53b0a980a40dee6f3d9332d92d933eea87

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js (hunk 1 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-11-misc/.decorator-evaluation-yield-super-private-all-private/exec.js]: This test is now enabled and merged with [decorator-evaluation-yield-private-super-property/exec.js](https://github.com/babel/babel/pull/16325/commits/707b91ce50f88b8f185d613c456bc1c4d600b287#diff-7dbf2473f4fc123ed19f0ce92ba8f3765e93a7a9fe856052b15e64f05a67f2e0)

It also happens to catch a bug in the `helper-replace-supers`, currently Babel is incorrectly replacing the `super` in the decorators of the given class method. This should have been handled in the `environmentVisitor` but the root class method path is not visited because the `visitSelf` argument of `traverse.node` is not exposed to the `path.traverse`.

https://github.com/babel/babel/pull/16325/commits/707b91ce50f88b8f185d613c456bc1c4d600b287

File: packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-exported/member-decorator/output.mjs [hunk 9]
```diff
@@ -1,7 +1,8 @@
 var _xDecs, _init_x;
 export class A {
+  static [(_xDecs = dec, "_")];
   static {
-    _xDecs = dec;
+    delete this._;
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2021-12-exported/member-decorator/output.mjs:5]: New PR idea: We can optimize them, potentially reusing the `usesFnContext` diagnostic info of each decorators sequence.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 10]
```diff
@@ -943,14 +1055,6 @@ function transformClass(
...
 
   if (!classDecorators && !hasElementDecorators) {
-    // If nothing is decorated but we have assignments, it must be the memoised
-    // computed keys of class accessors
-    if (assignments.length > 0) {
-      path.insertBefore(assignments.map(expr => t.expressionStatement(expr)));
-
-      // Recrawl the scope to make sure new identifiers are properly synced
-      path.scope.crawl();
-    }
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/decorators.ts:953]: This branch is removed because the computed key in a vanilla class accessor is now memoised in place.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 11]
```diff
@@ -1659,6 +1672,32 @@ function transformClass(
   }
 
   originalClass.body.body.unshift(t.staticBlock([]));
+  if (computedKeyAssignments.length > 0) {
+    // todo: the following branch will fail on 2023-05 version
+    if (version === "2023-11") {
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: This check is eventually removed in https://github.com/babel/babel/pull/16325/commits/bdee20d1461713f334b87bc53a5374fd05582d17.

**Comment by JLHwung** [comment]
I am investigating the test262 error.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 12]
```diff
@@ -754,12 +842,14 @@ function createPrivateBrandCheckClosure(brandName: t.PrivateName) {
...
 // `true` means "maybe" and `false` means "no".
-function usesFunctionContext(expression: t.Node) {
+function usesFunctionContextOrYieldAwait(expression: t.Node) {
   try {
     t.traverseFast(expression, node => {
       if (
         t.isThisExpression(node) ||
         t.isSuper(node) ||
+        t.isYieldExpression(node) ||
+        t.isAwaitExpression(node) ||
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/decorators.ts:852]: This function determines whether the class / element decorator evaluations should be moved into the `applyDecs` static block. If the decorator expression contains `await` or `yield`, we should not move it since the static block can not provide correct function context.

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-01-misc/all-decorators/output.js [hunk 13]
```diff
@@ -2,52 +2,50 @@ var _initProto, _initStatic, _initClass, _init_a, _init_d, _init_e, _call_f, _ca
...
-        this.#C = v;
-      }
+  static [class Class {
+    static {
+      ({
+        e: [_init_m, _call_o, _call_p, _call_q, _init_r, _get_r, _set_r, _init_d, _call_f, _call_g, _call_g2, _init_h, _get_h, _set_h, _init_i, _init_n, _init_a, _init_e, _initProto, _initStatic],
+        c: [_Class, _initClass]
+      } = babelHelpers.applyDecs2301(this, [[dec, 7, "j"], [dec, 8, "k"], [dec, 9, "l"], [dec, 6, "m"], [dec, 7, "o", function () {}], [dec, 8, "p", function () {}], [dec, 9, "q", function (v) {}], [dec, 6, "r", o => o.#D, (o, v) => o.#D = v], [dec, 2, "b"], [dec, 3, "c"], [dec, 4, "c"], [dec, 1, "d"], [dec, 2, "f", function () {}], [dec, 3, "g", function () {}], [dec, 4, "g", function (v) {}], [dec, 1, "h", o => o.#B, (o, v) => o.#B = v], [dec, 5, "i"], [dec, 5, "n", o => o.#n, (o, v) => o.#n = v], [dec, 0, "a"], [dec, 0, "e", o => o.#e, (o, v) => o.#e = v]], [dec], _ => #e in _));
+      _initStatic(this);
     }
```
Comment (liuxingbaoyu) [inline packages/babel-plugin-proposal-decorators/test/fixtures/2023-01-misc/all-decorators/output.js:12]: Maybe we can optimize this output in the future?

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 14]
```diff
@@ -1472,6 +1647,15 @@ function transformClass(
...
       path.replaceWith(newExpr);
+
+      path.traverse({
+        Class(path) {
+          if (path.node === originalClass) {
+            originalClassPath = path;
+            path.stop();
+          }
+        },
+      });
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/decorators.ts:1658]: We can directly access the `originalClass` path according to the AST hierarchy.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 15]
```diff
@@ -277,6 +278,93 @@ function extractProxyAccessorsFor(
...
 }
 
+function getComputedKeyCompletion(path: NodePath<t.Expression>) {
+  path = skipTransparentExprWrappers(path);
+  if (path.isSequenceExpression()) {
+    const expressions = path.get("expressions");
+    return getComputedKeyCompletion(expressions[expressions.length - 1]);
+  }
+  return path;
+}
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/decorators.ts:288]: Since this function is not meant for complex completion cases like `path.getCompletionRecord()` can handle, e.g. it does not support do expressions, I suggest rename it to `getComputedKeyLastElement` so that it is not confused with the completion concept in the spec.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts [hunk 16]
```diff
@@ -1493,36 +1677,86 @@ function transformClass(
...
-  originalClass.body.body.unshift(
-    t.staticBlock(
-      [
+
+  const applyDecoratorWrapper = t.staticBlock([]);
+  originalClass.body.body.unshift(applyDecoratorWrapper);
+  const applyDecsBody = applyDecoratorWrapper.body;
+  if (computedKeyAssignments.length > 0) {
+    const elements = originalClassPath.get("body.body");
+    let lastPublicElement: NodePath<t.ClassProperty | t.ClassMethod>;
```
Comment (JLHwung) [inline packages/babel-helper-create-class-features-plugin/src/decorators.ts:1686]: The `computedKeyAssignments` hosts decorator and computed keys evaluation results. They will be referenced in the `applyDecs` call emited in the first static block of the class. I think the output will be slightly more readable if we insert the assignments to the _first_ public element, so that they will not be too far away from where they will be used.

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts (hunk 14 above)
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: af8caa67c9c5f8e7a3e4a5dafdb453da29beea15

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts (hunk 15 above)
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts:299]: af8caa67c9c5f8e7a3e4a5dafdb453da29beea15

File: packages/babel-helper-create-class-features-plugin/src/decorators.ts (hunk 16 above)
Comment (JLHwung) [reply packages/babel-helper-create-class-features-plugin/src/decorators.ts]: https://github.com/babel/babel/pull/16325/commits/e856f3ccae87fbf5da14cdf41c12de34d6ef6910

File: packages/babel-plugin-proposal-decorators/test/fixtures/2023-01-misc/all-decorators/output.js (hunk 13 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-01-misc/all-decorators/output.js:12]: Sure. Though the output here is only one trailing `;` more than what we currently have. The output change here is introduced in https://github.com/babel/babel/pull/16325/commits/2d34eeb216b49f918ed0f34c279b9c65da9d4e52, which changes the output `static { class { ... } }` to `static [ class { ... } ];`.

Comment (liuxingbaoyu) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-01-misc/all-decorators/output.js:12]: I'm not sure if I can merge them into one class. :)
The changes here look good to me, the properties are kept in the outer class which greatly improves readability.

Comment (JLHwung) [reply packages/babel-plugin-proposal-decorators/test/fixtures/2023-01-misc/all-decorators/output.js:12]: > merge them into one class

The non static elements are split because the static fields have to be installed on the replaced class. If we can merge them without too much effort, https://github.com/babel/babel/issues/16176 will be automatically resolved.

**Review by nicolo-ribaudo — APPROVED** [review]
Well done!
